tests/integration/io/test_list_files_s3_minio.py

```diff
@@ -27,6 +27,199 @@ def s3fs_recursive_list(fs, path) -> list:
+                {"type": "File", "path": "s3://bucket/nested1/a.match", "size": 0},
+                {"type": "File", "path": "s3://bucket/nested2/a.match", "size": 0},
+            ],
+        ),
+        # Wildcard filepath after **
+        (
+            f"s3://bucket/**/nested1/*.match",
+            [
+                {"type": "File", "path": "s3://bucket/nested1/a.match", "size": 0},
+                {"type": "File", "path": "s3://bucket/nested1/c.match", "size": 0},
+            ],
+        ),
+        # Wildcard folder before **
+        (
+            f"s3://bucket/*/**/*.match",
+            [
+                {"type": "File", "path": "s3://bucket/nested1/a.match", "size": 0},
+                {"type": "File", "path": "s3://bucket/nested1/c.match", "size": 0},
+                {"type": "File", "path": "s3://bucket/nested2/a.match", "size": 0},
+                {"type": "File", "path": "s3://bucket/nested2/c.match", "size": 0},
+            ],
+        ),
+        ###
+        # `*`: wildcard
+        ###
+        # Wildcard file
+        (
+            f"s3://bucket/*.match",
+            [
+                {"type": "File", "path": "s3://bucket/a.match", "size": 0},
+                {"type": "File", "path": "s3://bucket/c.match", "size": 0},
+            ],
+        ),
+        # Nested wildcard file
+        (
+            f"s3://bucket/nested1/*.match",
+            [
+                {"type": "File", "path": "s3://bucket/nested1/a.match", "size": 0},
+                {"type": "File", "path": "s3://bucket/nested1/c.match", "size": 0},
+            ],
+        ),
+        # Wildcard folder + wildcard file
+        (
+            f"s3://bucket/*1/*.match",
+            [
+                {"type": "File", "path": "s3://bucket/nested1/a.match", "size": 0},
+                {"type": "File", "path": "s3://bucket/nested1/c.match", "size": 0},
+            ],
+        ),
+        # Wildcard folder + exact file
+        (
+            f"s3://bucket/*/a.match",
+            [
+                {"type": "File", "path": "s3://bucket/nested1/a.match", "size": 0},
+                {"type": "File", "path": "s3://bucket/nested2/a.match", "size": 0},
+            ],
+        ),
+        ###
+        # Missing paths
+        ###
+        # Exact filepath missing:
+        (f"s3://bucket/MISSING", []),
+        # Wildcard file no match:
+        (f"s3://bucket/*.MISSING", []),
+        # Exact folder missing:
+        (f"s3://bucket/MISSING/*.match", []),
+        # Wildcard folder no match:
+        (f"s3://bucket/*NOMATCH/*.match", []),
+        ###
+        # Directories: glob ignores directories and never returns them
+        ###
+        # Exact directory: we don't return directories
+        (f"s3://bucket/nested1", []),
+        # Wildcard folder: we don't select directories
+        (f"s3://bucket/nested*", []),
+        # Exact folder after **: we don't return directories
+        (f"s3://bucket/**/nested1", []),
```

Comment:
we should also add test cases for `s3://bucket/` and `s3://bucket/nested1` which should have the same functionality of `ls`

Reply:
See: https://github.com/Eventual-Inc/Daft/pull/1450#discussion_r1341771617